hrms/hr/doctype/full_and_final_statement/full_and_final_statement.py

```diff
@@ -8,6 +8,17 @@
 
 
 class FullandFinalStatement(Document):
+	def on_change(self):
+		for payable in self.payables:
+			if payable.component == "Gratuity":
+				if frappe.db.exists("Gratuity", payable.reference_document):
+					gratuity = frappe.get_doc("Gratuity", payable.reference_document)
+					if self.status == "Paid":
+						amount = payable.amount if self.docstatus == 1 else 0
+						gratuity.db_set("paid_amount", amount)
+					if self.docstatus == 2:
+						gratuity.set_status(cancel=True)
+
```

Comment:
1. Skip the db.exists call since doc deletion is not allowed if it has some linking
2. `on_change` will run on validate + deletion too. We just need this to run on `on_submit` and `on_cancel`
3. gratuity.set_status needs to be called in both the cases. On updating paid amount or cancelling FnF too
4. Reduce nesting in the function

Reply:
Thanks, Rucha!
Changes implemented:
- refactor code for updating status of gratuity records
-  set_status is called on linked gratuity docs whenever the FNF is canceled or its payment status is updated via the update_full_and_final_statement_status function.